vertx-service-discovery-backend-consul/src/main/java/io/vertx/servicediscovery/backend/consul/ConsulBackendService.java

```diff
@@ -0,0 +1,210 @@
+
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Objects;
+import java.util.UUID;
+import java.util.stream.Collectors;
+
+
+/**
+ * An implementation of the discovery backend based on Consul. Relies on {@link ConsulClient}.
+ *
+ * @author <a href="mailto:cafeinoman@openaliasbox.org>Francois Delalleau</a>
+ */
+public class ConsulBackendService implements ServiceDiscoveryBackend {
+
+  private ConsulClient client;
+
+  @Override
+  public void init(Vertx vertx, JsonObject config) {
+    ConsulClientOptions opt = new ConsulClientOptions(config);
+    this.client = ConsulClient.create(vertx, opt);
+  }
+
+  @Override
+  public void store(Record record, Handler<AsyncResult<Record>> resultHandler) {
+    String uuid = UUID.randomUUID().toString();
+    if (record.getRegistration() != null) {
+      throw new IllegalArgumentException("The record has already been registered");
+    }
+    ServiceOptions serviceOptions = recordToServiceOptions(record, uuid);
+    record.setRegistration(serviceOptions.getId());
+    Future<Void> registration = Future.future();
+    client.registerService(serviceOptions, registration);
+    registration.map(record).setHandler(resultHandler);
+  }
+
+  @Override
+  public void remove(Record record, Handler<AsyncResult<Record>> resultHandler) {
+    Objects.requireNonNull(record.getRegistration(), "No registration id in the record");
+    Future<Void> deregistration = Future.future();
+    client.deregisterService(record.getRegistration(), deregistration);
+    deregistration.map(record).setHandler(resultHandler);
+  }
+
+  @Override
+  public void remove(String uuid, Handler<AsyncResult<Record>> resultHandler) {
+    Objects.requireNonNull(uuid, "No registration id in the record");
+    getRecord(uuid, asyncRecord -> {
+      if (asyncRecord.succeeded()) {
+        remove(asyncRecord.result(), resultHandler);
+      } else {
+        resultHandler.handle(Future.failedFuture(asyncRecord.cause()));
+      }
+    });
+  }
+
+  @Override
+  public void update(Record record, Handler<AsyncResult<Void>> resultHandler) {
+    Objects.requireNonNull(record.getRegistration(), "No registration id in the record");
+    client.registerService(recordToServiceOptions(record, null), resultHandler);
+  }
+
+  @Override
+  public void getRecords(Handler<AsyncResult<List<Record>>> resultHandler) {
+    Future<ServiceList> nameList = Future.future();
+    client.catalogServices(nameList);
+    nameList.map(ServiceList::getList)
+      .map(l -> {
+        List<Future> recordFutureList = new ArrayList<>();
+        l.forEach(s -> {
+          if (!"consul".equals(s.getName())) {
+            ServiceQueryOptions opt = new ServiceQueryOptions();
+            if (!s.getTags().isEmpty()) {
+              opt.setTag(s.getTags().get(0));
+            }
+            Future<ServiceList> serviceList = Future.future();
+            client.catalogServiceNodesWithOptions(s.getName(), opt, serviceList);
```

Comment:
I agree, but I didn't use because I tought that we should return unhealthy node, as it is the reponsability of the ServiceDiscovery interface to filter the Record...